plugins/rest/src/server_command.cpp

```diff
@@ -0,0 +1,277 @@
+
+// FIXME: Don't use static variable for this.
+static bool s_require_authentication = true;
+
+static restinio::http_method_id_t to_restinio_method(vast::http_method method) {
+  switch (method) {
+    case vast::http_method::get:
+      return restinio::http_method_get();
+    case vast::http_method::post:
+      return restinio::http_method_post();
+  }
+  // Unreachable.
+  return restinio::http_method_get();
+}
+
+static caf::expected<enum server_config::mode>
+to_server_mode(const std::string& str) {
+  if (str == "debug")
+    return server_config::mode::debug;
+  else if (str == "upstream")
+    return server_config::mode::upstream;
+  else if (str == "server")
+    return server_config::mode::server;
+  else if (str == "mtls")
+    return server_config::mode::mtls;
+  return caf::make_error(ec::invalid_argument, "unknown mode");
+}
+
+static void
+setup_route(caf::scoped_actor& self, std::unique_ptr<router_t>& router,
+            std::string_view prefix,
+            const vast::rest_endpoint_plugin::api_endpoint& endpoint,
+            system::rest_handler_actor handler) {
+  auto method = to_restinio_method(endpoint.method);
+  auto path
+    = fmt::format("/api/v{}{}{}", static_cast<uint8_t>(endpoint.version),
+                  prefix, endpoint.path);
+  VAST_INFO("setting up route {}", path); // FIXME: INFO -> debug
+  router->add_handler(
+    method, path,
+    [&self, &endpoint,
+     handler](auto req, restinio::router::route_params_t /*route_params*/)
+      -> restinio::request_handling_status_t {
+      if (s_require_authentication) {
+        auto token = req->header().try_get_field("X-VAST-Token");
+        if (!token) {
+          req->create_response(restinio::status_unauthorized())
+            .set_body("missing `X-VAST-Token` header")
+            .done();
+          return restinio::request_rejected();
+        }
+        VAST_INFO("got token: {}", *token); // FIXME: remove msg
+        auto& server = server_singleton();
+        if (server.authenticate(*token)) {
+          req->create_response(restinio::status_unauthorized())
+            .set_body("invalid token")
+            .done();
+          return restinio::request_rejected();
+        }
+      }
+      restinio::query_string_params_t string_params
+        = restinio::parse_query(req->header().query());
+      VAST_INFO("handling request with {}  parameters", string_params.size());
+      // TODO: warn about unknown/unexpected paramers
+      // for (auto& value : string_params) {
+      // }
+      vast::record params;
+      if (endpoint.params) {
+        for (auto const& leaf : endpoint.params->leaves()) {
+          auto name = leaf.field.name;
+          auto restinio_name
+            = restinio::string_view_t{name.data(), name.size()};
+          VAST_INFO("checking request parameter {}", name);
+          if (string_params.has(restinio_name)) {
+            auto value = std::string{string_params[restinio_name]};
```

Comment:
@dominiklohmann Here we still need to cast value to the type expected to the function.

Reply:
This is fixed for at least basic types, but there's still an outstanding TODO for lists. Did you want to address that still, or is that waiting for the story that is about implementing "parse this string as a data with this type information" more generically?

Reply:
I'll probably create a separate story for supporting lists, so we can implement it when we actually want to add the first endpoint with a list parameter.